src/libsync/propagatedownload.cpp

```diff
@@ -821,8 +827,26 @@ void PropagateDownloadFile::slotGetFinished()
     validator->start(_tmpFile.fileName(), checksumHeader);
 }
 
-void PropagateDownloadFile::slotChecksumFail(const QString &errMsg)
+void PropagateDownloadFile::slotChecksumFail(const QString &errMsg, const QByteArray &checksumType, const QByteArray &checksum, const QString &filePath)
 {
+    if (!checksumType.isEmpty() && !checksum.isEmpty() && !filePath.isEmpty()) {
+        ConfigFile cfgFile;
+
+        if (cfgFile.allowChecksumValidationFail()) {
+            const QString key = checksum + _item->_fileId;
+            const auto numChecksumMismatchEntries = propagator()->_journal->keyValueStoreGetInt(QString(key), 0);
+            if (numChecksumMismatchEntries < numChecksumFailuresAllowed) {
+                qCWarning(lcPropagateDownload) << "Checksum validation has failed" << numChecksumMismatchEntries << " times, but, allowChecksumValidationFail is set.Let's give it another try...";
+                propagator()->_journal->keyValueStoreSet(QString(key), numChecksumMismatchEntries + 1);
+            } else {
+                propagator()->_journal->keyValueStoreDelete(QString(key));
+                qCWarning(lcPropagateDownload) << "Checksum validation has failed" << numChecksumMismatchEntries << " times, but, allowChecksumValidationFail is set, so, let's continue...";
```

Comment:
And there I would probably say that we ignore checksum mismatch after `numChecksumMismatchEntries` retry where the downloaded file would always have another checksum value (different from the expected one). I would also keep it info because we are indeed taking a decision but we never now if this is an error or just a checksum that is not up to date.

Reply:
@mgallien But, why is the checksum not up-to-date? Isn't this erroneous behavior? It doesn't happen with other users, right? We are just facing it with one user, so, it is probably something that does not normally happen?

Reply:
if I understand what Robin was saying the server can only do a best effort to ensure that checksum is invalidated when a file is modified but cannot guarantee it
your user may have it now but you cannot exclude that other users may also encounter this situation
we should build something that is working in general and that we can keep in our code for a long time to help those situations

Reply:
> if I understand what Robin was saying the server can only do a best effort to ensure that checksum is invalidated when a file is modified but cannot guarantee it
> your user may have it now but you cannot exclude that other users may also encounter this situation
> we should build something that is working in general and that we can keep in our code for a long time to help those situations

@mgallien Yes, I understand that, that's why I haven't proposed to check for specific user's server host hash but, apply this logic to all users even those that don't experience this issue now but will in the future. Yet, I am not completely on the same page with you about turning this warning into info. We have both info and warning written into a log and, this scenario still doesn't happen to the majority of users. But, I am going to change it to info, as, it is not something I want to fight for :)